spring-rabbit/src/main/java/org/springframework/amqp/rabbit/annotation/RabbitListenerAnnotationBeanPostProcessor.java

```diff
@@ -497,37 +510,100 @@ private String resolve(String value) {
-							resolveExpressionAsBoolean(bindingExchange.autoDelete()));
-					actualBinding = new Binding(queueName, DestinationType.QUEUE, exchangeName, resolvedKey, null);
+							resolveExpressionAsBoolean(bindingExchange.autoDelete()),
+							resolveArguments(bindingExchange.arguments()));
+					actualBinding = new Binding(queueName, DestinationType.QUEUE, exchangeName, resolvedKey,
+							resolveArguments(binding.arguments()));
 				}
 				else if (exchangeType.equals(ExchangeTypes.FANOUT)) {
 					exchange = new FanoutExchange(exchangeName,
 							resolveExpressionAsBoolean(bindingExchange.durable()),
-							resolveExpressionAsBoolean(bindingExchange.autoDelete()));
-					actualBinding = new Binding(queueName, DestinationType.QUEUE, exchangeName, "", null);
+							resolveExpressionAsBoolean(bindingExchange.autoDelete()),
+							resolveArguments(bindingExchange.arguments()));
+					actualBinding = new Binding(queueName, DestinationType.QUEUE, exchangeName, "",
+							resolveArguments(binding.arguments()));
 				}
 				else if (exchangeType.equals(ExchangeTypes.TOPIC)) {
 					exchange = new TopicExchange(exchangeName,
 							resolveExpressionAsBoolean(bindingExchange.durable()),
-							resolveExpressionAsBoolean(bindingExchange.autoDelete()));
-					actualBinding = new Binding(queueName, DestinationType.QUEUE, exchangeName, resolvedKey, null);
+							resolveExpressionAsBoolean(bindingExchange.autoDelete()),
+							resolveArguments(bindingExchange.arguments()));
+					actualBinding = new Binding(queueName, DestinationType.QUEUE, exchangeName, resolvedKey,
+							resolveArguments(binding.arguments()));
+				}
+				else if (exchangeType.equals(ExchangeTypes.HEADERS)) {
+					exchange = new HeadersExchange(exchangeName,
+							resolveExpressionAsBoolean(bindingExchange.durable()),
+							resolveExpressionAsBoolean(bindingExchange.autoDelete()),
+							resolveArguments(bindingExchange.arguments()));
+					actualBinding = new Binding(queueName, DestinationType.QUEUE, exchangeName, resolvedKey,
+							resolveArguments(binding.arguments()));
 				}
 				else {
 					throw new BeanInitializationException("Unexpected exchange type: " + exchangeType);
 				}
-				((ConfigurableBeanFactory) this.beanFactory).registerSingleton(exchangeName + ++this.increment, exchange);
-				((ConfigurableBeanFactory) this.beanFactory).registerSingleton(exchangeName + ++this.increment, actualBinding);
+				((ConfigurableBeanFactory) this.beanFactory).registerSingleton(exchangeName + ++this.increment,
+						exchange);
+				((ConfigurableBeanFactory) this.beanFactory).registerSingleton(exchangeName + ++this.increment,
+						actualBinding);
 			}
 		}
 		return queues.toArray(new String[queues.size()]);
 	}
 
+	private Map<String, Object> resolveArguments(Argument[] arguments) {
+		Map<String, Object> map = new HashMap<String, Object>();
+		for (Argument arg : arguments) {
+			String key = resolve(arg.name());
+			if (StringUtils.hasText(key)) {
```

Comment:
In pursuit of the `@Argument(name)` property, I think we shouldn't allow to resolve it to the empty `String`.
Although... it may make sense in some cases when we deploy application into different environment. 
WDYT?